Created MOOSE Meshing Tutorial documentation (#24041)
idaholab/moose: modules/doc/content/getting_started/examples_and_tutorials/tutorial04_meshing/step14_advanced_examples.md

!media tutorial04_meshing/adv_ex_2DMSRE.png
       id=tutorial04-adv_ex_2DMSRE
       caption=Molten Salt Reactor Experiment (MSRE) 2D lattice.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

### KRUSTY heat-pipe cooled microreactor 2D core

- [ParsedCurveGenerator.md] is used for creating the outline of complex shapes before mesh generation.
- [XYDelaunayGenerator.md] is employed for meshing the irregular areas.
- [PeripheralTriangleMeshGenerator.md] is utilized to add layers of triangular mesh onto an existing circular mesh.
- !alert construction
  CircularCorrectionGenerator, which is currently under development, is used to adjust a circle to preserve its volume.

!media tutorial04_meshing/adv_ex_2DKRUSTY.png
       id=tutorial04-adv_ex_2DKRUSTY
       caption=KRUSTY heat pipe-cooled microreactor 2D core.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

### Modular High Temperature Gas Cooled Reactor (MHTGR)

- [PolygonConcentricCircleMeshGenerator.md] and [PatternedHexMeshGenerator.md] generate pins and assemblies
- [XYDelaunayGenerator.md] generates control rod holes and reflector assemblies

> - [StitchedMeshGenerator.md] merges control rod holes with fuel g reflector assemblies

Reviewer: ```suggestion
- [StitchedMeshGenerator.md] merges control rod holes with fuel and reflector assemblies
```

- [HexagonConcentricCircleAdaptiveBoundaryMeshGenerator.md] generates ducts for reflector assembly that can be stitched with other assembles.

> - [PatternedHexMeshGenerator.md] merges fuel and reflector assemblies into the core

Reviewer: ```suggestion
- [PatternedHexMeshGenerator.md] generates a lattice of fuel and reflector assemblies into the core
```

> - [PeripheralRingMeshGenerator.md] adds circular boundary surrounding the core

Reviewer: ```suggestion
- [PeripheralRingMeshGenerator.md] adds a circular boundary surrounding the core
```

- [AdvancedExtruderGenerator.md] performs 2D to 3D extrusion for the core
- [DepletionIDGenerator.md] sets up assembly-wise depletion zones

!media tutorial04_meshing/adv_ex_mhgtr.png
       id=tutorial04-adv_ex_mhgtr
       caption=MHTGR 2D mesh.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!media tutorial04_meshing/adv_ex_mhtgrzoom.png
       id=tutorial04-adv_ex_mhtgrzoom
       caption=MHTGR 2D mesh, zoomed in.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

### 2D Pebble Bed Reactor with Streamlines

- [FillBetweenPointVectorsGenerator.md] generates pebble bed streamlines
- [CartesianMeshGenerator.md] and [FillBetweenSidesetsGenerator.md] generates rectangular meshes and trapezoid meshes

!media tutorial04_meshing/adv_ex_PBRPronghornGriffin.png
       id=tutorial04-adv_ex_PBRPronghornGriffin
       caption=2D generic pebble bed reactor meshes for Pronghorn (left) and Griffin (right) with flow streamlines from top inlet to bottom conical outlet.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

> ## Advanced Meshing Examples: Boundary Layers and Biasing

Reviewer 2: ```suggestion
## Boundary Layers and Biasing
```

       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!listing sec14_adv/rad_bias_prmg.i
         id=tutorial04-rad_bias_prmg.i
         caption=Peripheral ring radial mesh biasing example.
         block=Mesh

### Mesh Boundary Layers

Boundary layers are commonly needed in thermal-hydraulic applications, where the thickness of the mesh near a wall must be a certain size, uniformly, regardless of the width of the meshed region. Boundary layers are possible using [PolygonConcentricCircleMeshGenerator.md] and [PeripheralRingMeshGenerator.md].

!media tutorial04_meshing/pccmg_bdry_layer.png
       id=tutorial04-pccmg_bdry_layer
       caption=A hexagonal pin mesh generated by [PolygonConcentricCircleMeshGenerator.md] with boundary layers created in both the outmost ring block (outer boundary layer, red) and background block (inner boundary layer, blue).
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!listing sec14_adv/bdry_layer_pccmg.i
         id=tutorial04-bdry_layer_pccmg.i
         caption=[PolygonConcentricCircleMeshGenerator.md] boundary layer example.
         block=Mesh

> ## Advanced Meshing Example: Stitching Assemblies with Different Pin Numbers

Reviewer 2: ```suggestion
## Stitching Assemblies with Different Pin Numbers
```

The best approach for this situation is to use [PatternedHexPeripheralModifier.md] to strip the outermost layer off each assembly and replace it with a transition layer with a set number of nodes on the outermost boundary. Doing this to each assembly will ensure that all assemblies have the same number of boundary nodes and can be easily stitched. An example follows.

!media tutorial04_meshing/adv_ex_assembies_with_different_pin_nums.png
       id=tutorial04-adv_ex_assembies_with_different_pin_nums
       caption=[PatternedHexPeripheralModifier.md] is used to adjust the boundary node count on each assembly to be consistent with each other and therefore stitchable.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!listing modules/reactor/test/tests/meshgenerators/patterned_hex_peripheral_modifier/patterned.i
         id=tutorial04-patterned.i
         caption=Assemblies with different pin numbers example.
         block=Mesh

> ## Advanced Meshing Example: Oversized Pin

Reviewer 2: ```suggestion
## Oversized Pin
```

Separately, the oversized pin should be defined as its own object, remembering to remove the background region which is created by default in [PolygonConcentricCircleMeshGenerator.md]. This mesh will serve as the "hole" for [XYDelaunayGenerator.md].

Finally, [XYDelaunayGenerator.md] is used to mesh the regions between the assembly (with deleted dummies) and the oversized pin. This is done by giving [XYDelaunayGenerator.md] the assembly mesh and the oversized pin mesh and specifying via input parameters that the region between the assembly mesh and the oversized pin mesh should be triangulated. The oversized pin mesh is placed into the geometry at this time as well.

!media tutorial04_meshing/adv_ex_oversized_pin.png
       id=tutorial04-adv_ex_oversized_pin
       caption=An assembly mesh with an oversized pin in the center. The mesh is generated by modifying the output of [PatternedHexMeshGenerator.md] by replacing the central region with a mesh generated by [XYDelaunayGenerator.md].
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!listing sec14_adv/oversize.i
         id=tutorial04-oversize.i
         caption=Oversized pin example.
         block=Mesh

> ## Advanced Meshing Example: Filling Between Curves Example of using "FillBetweenSidesetsGenerator"

Reviewer 2: ```suggestion
## Filling Between Curves using "FillBetweenSidesetsGenerator"
```

!media tutorial04_meshing/adv_ex_fbsg.png
       id=tutorial04-adv_ex_fbsg
       caption=An example of using [FillBetweenSidesetsGenerator.md] to connect quarter circular-shapedsquare meshes.
       style=width:50%;display:block;margin-left:auto;margin-right:auto;

!listing sec14_adv/quarter_circle_connect.i
         id=tutorial04-quarter_circle_connect.i
         caption=[FillBetweenSidesetsGenerator.md] example.
         block=Mesh/connect_two_circles

> ## Advanced Meshing Example: Triangulation of Odd-Shaped Regions Using "ParsedCurveGenerator" and "XYDelaunayGenerator"

Reviewer 2: ```suggestion
## Triangulation of Odd-Shaped Regions Using "ParsedCurveGenerator" and "XYDelaunayGenerator"
```
